File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/bridge/functions/EdgeFunctions.kt
package com.nativephp.mobile.bridge.functions

import android.util.Log
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.ui.NativeUIState
import kotlinx.coroutines.launch

private const val TAG = "EdgeFunctions"

/**
 * Functions related to Edge UI management
 * Namespace: "Edge.*"
 */
object EdgeFunctions {

    /**
     * Update the native UI state with Edge components
     * Parameters:
     *   - components: array - Array of Edge components
     *
     * Usage Example:
     *   nativephp_call('Edge.Set', json_encode([
     *     'components' => [
     *       ['type' => 'bottom_nav', 'data' => [...]]
     *     ]
     *   ]));
     */
    class Set : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            // Extract components from parameters
            val components = parameters["components"]

            if (components == null) {
                Log.e(TAG, "❌ No components provided in parameters")
                return mapOf("error" to "No components provided")
            }

            Log.d(TAG, "🎨 Edge.Set called")

            try {
                // Pass the components object directly to NativeUIState
                // It will handle the type conversion internally
                NativeUIState.updateFromJson(components)

                return mapOf("success" to true)
            } catch (e: Exception) {
                Log.e(TAG, "❌ Error updating UI state: ${e.message}", e)
                return mapOf("error" to "Failed to update UI state: ${e.message}")
            }
        }
    }

    /**
     * Programmatically open the side navigation drawer
     *
     * Usage Example:
     *   nativephp_call('Navigation.OpenSidebar', '{}');
     */
    class OpenSidebar : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Log.d(TAG, "🎨 Navigation.OpenSidebar called")

            // Check if side nav data exists
            if (!NativeUIState.hasSideNav()) {
                Log.w(TAG, "⚠️ Navigation.OpenSidebar: No side nav data available")
                return mapOf("success" to false, "error" to "No side nav data available")
            }

            val scope = NativeUIState.drawerScope
            val state = NativeUIState.drawerState

            if (scope == null || state == null) {
                Log.e(TAG, "❌ Navigation.OpenSidebar: Drawer state not initialized (scope=$scope, state=$state)")
                return mapOf("success" to false, "error" to "Drawer state not initialized")
            }

            scope.launch {
                state.open()
                Log.d(TAG, "✅ Navigation.OpenSidebar: Sidebar opened")
            }

            return mapOf("success" to true)
        }
    }

    /**
     * Programmatically close the side navigation drawer
     *
     * Usage Example:
     *   nativephp_call('Navigation.CloseSidebar', '{}');
     */
    class CloseSidebar : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Log.d(TAG, "🎨 Navigation.CloseSidebar called")

            val scope = NativeUIState.drawerScope
            val state = NativeUIState.drawerState

            if (scope == null || state == null) {
                Log.e(TAG, "❌ Navigation.CloseSidebar: Drawer state not initialized (scope=$scope, state=$state)")
                return mapOf("success" to false, "error" to "Drawer state not initialized")
            }

            scope.launch {
                state.close()
                Log.d(TAG, "✅ Navigation.CloseSidebar: Sidebar closed")
            }

            return mapOf("success" to true)
        }
    }

    /**
     * Toggle the side navigation drawer open/closed state
     *
     * Usage Example:
     *   nativephp_call('Navigation.ToggleSidebar', '{}');
     */
    class ToggleSidebar : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Log.d(TAG, "🎨 Navigation.ToggleSidebar called")

            // Check if side nav data exists
            if (!NativeUIState.hasSideNav()) {
                Log.w(TAG, "⚠️ Navigation.ToggleSidebar: No side nav data available")
                return mapOf("success" to false, "error" to "No side nav data available")
            }

            val scope = NativeUIState.drawerScope
            val state = NativeUIState.drawerState

            if (scope == null || state == null) {
                Log.e(TAG, "❌ Navigation.ToggleSidebar: Drawer state not initialized (scope=$scope, state=$state)")
                return mapOf("success" to false, "error" to "Drawer state not initialized")
            }

            scope.launch {
                if (state.isOpen) {
                    state.close()
                    Log.d(TAG, "✅ Navigation.ToggleSidebar: Sidebar closed")
                } else {
                    state.open()
                    Log.d(TAG, "✅ Navigation.ToggleSidebar: Sidebar opened")
                }
            }

            return mapOf("success" to true)
        }
    }
}

File: resources/xcode/NativePHP/Bridge/Functions/EdgeFunctions.swift
import Foundation
import SwiftUI

// MARK: - Edge Function Namespace

/// Functions related to Edge UI management
/// Namespace: "Edge.*"
enum EdgeFunctions {

    // MARK: - Edge.Set

    /// Update the native UI state with Edge components
    /// Parameters:
    ///   - components: array - Array of Edge components
    ///
    /// Usage Example:
    ///   nativephp_call('Edge.Set', json_encode([
    ///     'components' => [
    ///       ['type' => 'bottom_nav', 'data' => [...]]
    ///     ]
    ///   ]));
    class Set: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            // Extract components array from parameters
            guard let components = parameters["components"] as? [[String: Any]] else {
                print("❌ Edge.Set: No components array provided")
                return ["error": "No components array provided"]
            }

            print("🎨 Edge.Set called with \(components.count) component(s)")
            print("🎨 Edge.Set components: \(components)")

            // Convert components back to JSON string for NativeUIState
            do {
                let jsonData = try JSONSerialization.data(withJSONObject: components, options: [])
                guard let jsonString = String(data: jsonData, encoding: .utf8) else {
                    print("❌ Edge.Set: Failed to convert components to JSON string")
                    return ["error": "Failed to convert components to JSON string"]
                }

                // Update NativeUIState on main thread synchronously
                // Use sync to ensure the UI state is updated before PHP response completes
                // This prevents a race condition where the WebView receives the response
                // before the EDGE components are registered in NativeUIState
                if Thread.isMainThread {
                    NativeUIState.shared.updateFromJson(jsonString)
                } else {
                    DispatchQueue.main.sync {
                        NativeUIState.shared.updateFromJson(jsonString)
                    }
                }

                return ["success": true]
            } catch {
                print("❌ Edge.Set: JSON serialization error: \(error)")
                return ["error": "JSON serialization failed: \(error.localizedDescription)"]
            }
        }
    }

    // MARK: - Navigation.OpenSidebar

    /// Programmatically open the side navigation drawer
    ///
    /// Usage Example:
    ///   nativephp_call('Navigation.OpenSidebar', '{}');
    class OpenSidebar: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            print("🎨 Navigation.OpenSidebar called")

            // Check if side nav exists
            guard NativeUIState.shared.hasSideNav() else {
                print("⚠️ Navigation.OpenSidebar: No side nav data available")
                return ["success": false, "error": "No side nav data available"]
            }

            if Thread.isMainThread {
                withAnimation(.easeInOut(duration: 0.3)) {
                    NativeUIState.shared.openSidebar()
                }
            } else {
                DispatchQueue.main.sync {
                    withAnimation(.easeInOut(duration: 0.3)) {
                        NativeUIState.shared.openSidebar()
                    }
                }
            }

            print("✅ Navigation.OpenSidebar: Sidebar opened")
            return ["success": true]
        }
    }

    // MARK: - Navigation.CloseSidebar

    /// Programmatically close the side navigation drawer
    ///
    /// Usage Example:
    ///   nativephp_call('Navigation.CloseSidebar', '{}');
    class CloseSidebar: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            print("🎨 Navigation.CloseSidebar called")

            if Thread.isMainThread {
                withAnimation(.easeInOut(duration: 0.3)) {
                    NativeUIState.shared.closeSidebar()
                }
            } else {
                DispatchQueue.main.sync {
                    withAnimation(.easeInOut(duration: 0.3)) {
                        NativeUIState.shared.closeSidebar()
                    }
                }
            }

            print("✅ Navigation.CloseSidebar: Sidebar closed")
            return ["success": true]
        }
    }

    // MARK: - Navigation.ToggleSidebar

    /// Toggle the side navigation drawer open/closed state
    ///
    /// Usage Example:
    ///   nativephp_call('Navigation.ToggleSidebar', '{}');
    class ToggleSidebar: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            print("🎨 Navigation.ToggleSidebar called")

            // Check if side nav exists
            guard NativeUIState.shared.hasSideNav() else {
                print("⚠️ Navigation.ToggleSidebar: No side nav data available")
                return ["success": false, "error": "No side nav data available"]
            }

            let willOpen = !NativeUIState.shared.shouldPresentSidebar

            if Thread.isMainThread {
                withAnimation(.easeInOut(duration: 0.3)) {
                    NativeUIState.shared.shouldPresentSidebar.toggle()
                }
            } else {
                DispatchQueue.main.sync {
                    withAnimation(.easeInOut(duration: 0.3)) {
                        NativeUIState.shared.shouldPresentSidebar.toggle()
                    }
                }
            }

            print("✅ Navigation.ToggleSidebar: Sidebar \(willOpen ? "opened" : "closed")")
            return ["success": true]
        }
    }
}